src/persist/src/indexed/background.rs

```diff
@@ -154,68 +155,160 @@ impl<B: Blob> Maintainer<B> {
+            if first_updates.is_empty() {
+                if let Some(first) = first_batch_iter.next() {
+                    let first_batch = blob
+                        .get_trace_batch_async(first, CacheHint::NeverAdd)
+                        .recv()?;
+                    first_updates.extend(first_batch.updates.iter().flat_map(|u| {
+                        u.iter()
+                            .map(|((k, v), t, d)| ((k.to_vec(), v.to_vec()), t, d))
+                    }));
+                }
+            }
+
+            if second_updates.is_empty() {
+                if let Some(second) = second_batch_iter.next() {
+                    let second_batch = blob
+                        .get_trace_batch_async(second, CacheHint::NeverAdd)
+                        .recv()?;
+                    second_updates.extend(second_batch.updates.iter().flat_map(|u| {
+                        u.iter()
+                            .map(|((k, v), t, d)| ((k.to_vec(), v.to_vec()), t, d))
+                    }));
+                }
+            }
+
+            // Figure out the stopping point for records we can safely consolidate.
+            // This is the min max value across the two batches we have seen so far.
+            let threshold_key_val = match (first_updates.last(), second_updates.last()) {
+                (Some((f, _, _)), Some((s, _, _))) => {
+                    if f < s {
+                        f.clone()
+                    } else {
+                        s.clone()
+                    }
+                }
+                (Some((f, _, _)), None) => f.clone(),
+                (None, Some((s, _, _))) => s.clone(),
+                (None, None) => break,
+            };
+
+            // Drain all of the records <= threshold out of the various holding
+            // pens.
+            //
+            // TODO: this could be replaced with a drain_filter if that wasn't
+            // experimental.
+            let mut i = 0;
+            while i < first_updates.len() {
+                if first_updates[i].0 <= threshold_key_val {
+                    let data = first_updates.remove(i);
+                    updates.push(data);
+                } else {
+                    break;
+                }
+            }
+            i = 0;
+            while i < second_updates.len() {
+                if second_updates[i].0 <= threshold_key_val {
+                    let data = second_updates.remove(i);
+                    updates.push(data);
+                } else {
+                    break;
+                }
+            }
+
+            for ((_, _), ts, _) in updates.iter_mut() {
+                ts.advance_by(desc.since().borrow());
+            }
+            differential_dataflow::consolidation::consolidate_updates(&mut updates);
+
+            for ((key, val), time, diff) in updates.iter() {
+                builder.push(((key, val), *time, *diff));
+                if builder.num_filled() > 0 {
```

Comment:
instead of num_filled/take_filled, why not only take_filled but it returns an Option?

Reply:
done.